internal/gatewayapi/securitypolicy.go

```diff
@@ -316,23 +385,122 @@ func (t *Translator) buildCORS(policy *egv1a1.SecurityPolicy) *ir.CORS {
 				Suffix: &origin.Value,
 			})
 		case egv1a1.StringMatchRegularExpression:
+			if err := validateRegex(origin.Value); err != nil {
+				return nil, err // TODO zhaohuabing: also check regex in other places
+			}
 			allowOrigins = append(allowOrigins, &ir.StringMatch{
-				SafeRegex: &origin.Value, // TODO zhaohuabing: check if the value is a valid regex
+				SafeRegex: &origin.Value,
 			})
 		}
 	}
 
 	return &ir.CORS{
 		AllowOrigins:  allowOrigins,
-		AllowMethods:  policy.Spec.CORS.AllowMethods,
-		AllowHeaders:  policy.Spec.CORS.AllowHeaders,
-		ExposeHeaders: policy.Spec.CORS.ExposeHeaders,
-		MaxAge:        policy.Spec.CORS.MaxAge,
-	}
+		AllowMethods:  cors.AllowMethods,
+		AllowHeaders:  cors.AllowHeaders,
+		ExposeHeaders: cors.ExposeHeaders,
+		MaxAge:        cors.MaxAge,
+	}, nil
 }
 
-func (t *Translator) buildJWT(policy *egv1a1.SecurityPolicy) *ir.JWT {
+func (t *Translator) buildJWT(jwt *egv1a1.JWT) *ir.JWT {
 	return &ir.JWT{
-		Providers: policy.Spec.JWT.Providers,
+		Providers: jwt.Providers,
+	}
+}
+
+func (t *Translator) buildOIDC(
+	policy *egv1a1.SecurityPolicy,
+	resources *Resources) (*ir.OIDC, error) {
+	var (
+		oidc         = policy.Spec.OIDC
+		clientSecret *v1.Secret
+		err          error
+	)
+
+	from := crossNamespaceFrom{
+		group:     egv1a1.GroupName,
+		kind:      KindSecurityPolicy,
+		namespace: policy.Namespace,
 	}
+	if clientSecret, err = t.validateSecretRef(from, oidc.ClientSecret, resources); err != nil {
+		return nil, err
+	}
+
+	val, ok := clientSecret.Data[egv1a1.OIDCClientSecretKey]
+	if !ok || len(val) == 0 {
+		return nil, fmt.Errorf("client secret not found in secret %s/%s", clientSecret.Namespace, clientSecret.Name)
+	}
+
+	// Discover the token and authorization endpoints from the issuer's
+	// well-known url if not explicitly specified
+	provider := oidc.Provider.DeepCopy()
```

Comment:
curious why deep copy is needed here

Reply:
Because the discovered token and authorization endpoints will be added to the copied provider.